crates/red_knot_test/src/parser.rs

```diff
@@ -0,0 +1,547 @@
+        let mut parent_id = self.section.parent_id;
+        while let Some(next_id) = parent_id {
+            let parent = &self.suite.sections[next_id];
+            parent_id = parent.parent_id;
+            if !name.is_empty() {
+                name.insert_str(0, " - ");
+            }
+            name.insert_str(0, parent.title);
+        }
+        if !name.is_empty() {
+            name.push_str(" - ");
+        }
+        name.push_str(self.section.title);
+        name
+    }
+
+    pub(crate) fn files(&self) -> impl Iterator<Item = &'m EmbeddedFile<'s>> {
+        self.files.iter()
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MarkdownTestIterator<'m, 's> {
+    suite: &'m MarkdownTestSuite<'s>,
+    current_file_index: usize,
+}
+
+impl<'m, 's> Iterator for MarkdownTestIterator<'m, 's> {
+    type Item = MarkdownTest<'m, 's>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut current_file_index = self.current_file_index;
+        let mut file = self.suite.files.get(current_file_index.into());
+        let section_id = file?.section;
+        while file.is_some_and(|file| file.section == section_id) {
+            current_file_index += 1;
+            file = self.suite.files.get(current_file_index.into());
+        }
+        let files = &self.suite.files[EmbeddedFileId::from_usize(self.current_file_index)
+            ..EmbeddedFileId::from_usize(current_file_index)];
+        self.current_file_index = current_file_index;
+        Some(MarkdownTest {
+            suite: self.suite,
+            section: &self.suite.sections[section_id],
+            files,
+        })
+    }
+}
+
+#[newtype_index]
+struct SectionId;
+
+#[derive(Debug)]
+struct Section<'s> {
+    title: &'s str,
+    level: u8,
+    parent_id: Option<SectionId>,
+}
+
+#[newtype_index]
+struct EmbeddedFileId;
+
+#[derive(Debug)]
+pub(crate) struct EmbeddedFile<'s> {
+    section: SectionId,
+    pub(crate) path: &'s str,
+    pub(crate) lang: &'s str,
+    pub(crate) code: &'s str,
+}
+
+/// Matches an arbitrary amount of whitespace (including newlines), followed by a sequence of `#`
+/// characters, followed by a title heading, followed by a newline.
+static HEADER_RE: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"^(\s*\n)*(?<level>#+)\s+(?<title>.+)\s*\n").unwrap());
+
+/// Matches a code block fenced by triple backticks, possibly with language and `key=val`
+/// configuration items following the opening backticks (in the "tag string" of the code block).
+static CODE_RE: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(r"^```(?<lang>\w+)(?<config>( +\S+)*)\s*\n(?<code>(.|\n)*?)\n```\s*\n").unwrap()
+});
+
+#[derive(Debug)]
+struct SectionStack(Vec<SectionId>);
+
+impl SectionStack {
+    fn new(root_section_id: SectionId) -> Self {
+        Self(vec![root_section_id])
+    }
+
+    fn push(&mut self, section_id: SectionId) {
+        self.0.push(section_id);
+    }
+
+    fn pop(&mut self) -> Option<SectionId> {
+        self.0.pop()
+    }
```

Comment:
Does this have the same invariant as `parent()` below? Is it worth asserting that invariant in debug builds?

```suggestion
    fn pop(&mut self) {
        let popped = self.0.pop();
        debug_assert_ne!(popped, None, "Should never pop the implicit root section");
    }
```